File: LinkList/leetcode/07.go
package leetcode

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func getIntersectionNode(headA, headB *ListNode) *ListNode {
	curA := headA
	curB := headB
	lenA, lenB := 0, 0
	// 求A，B的长度
	for curA != nil {
			curA = curA.Next
			lenA++
	}
	for curB != nil {
			curB = curB.Next
			lenB++
	}
	var step int
	var fast, slow *ListNode
	// 请求长度差，并且让更长的链表先走相差的长度
	if lenA > lenB {
			step = lenA - lenB
			fast, slow = headA, headB
	} else {
			step = lenB - lenA
			fast, slow = headB, headA
	}
	for i:=0; i < step; i++ {
			fast = fast.Next
	}
	// 遍历两个链表遇到相同则跳出遍历
	for fast != slow {
			fast = fast.Next
			slow = slow.Next
	}
	return fast
}

File: LinkList/cpp/07.cpp
#include "head/listnode.h"

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode* curA = headA;
        ListNode* curB = headB;
        int lenA = 0, lenB = 0;
        while (curA != nullptr) { // 求链表A的长度
            lenA++;
            curA = curA->next;
        }
        while (curB != nullptr) { // 求链表B的长度
            lenB++;
            curB = curB->next;
        }
        curA = headA;
        curB = headB;
        ListNode* fast;
        ListNode* slow;
        int gap;

        if (lenA > lenB) {
            gap = lenA-lenB;
            slow = curB;
            fast = curA;
        } else {
            gap = lenB-lenA;
            slow = curA;
            fast = curB;
        }

        for (int i = 0; i < gap; i++) {
            fast = fast->next;
        }

        while(fast != slow) {
            fast = fast->next;
            slow = slow->next;
        }
        return fast;
    }
};